mitiq/folding_cirq.py

```diff
@@ -196,11 +196,28 @@ def fold_gates_at_random(circuit: Circuit, stretch: float, seed: Optional[int])
     ngates = len(list(folded.all_operations()))
     num_to_fold = int(ngates * (stretch - 1.0) / 2.0)
 
+    # Keep track of where moments are in the folded circuit
+    moment_indices = {i: i for i in range(len(circuit))}
+
+    # Keep track of which gates we can fold in each moment
+    remaining_gate_indices = {moment: list(range(len(circuit[moment]))) for moment in range(len(circuit))}
+
     for _ in range(num_to_fold):
-        # TODO: Update gate choices properly using tracked moment indices.
-        moment_index = np.random.choice(len(folded))
-        gate_index = np.random.choice(len(folded[moment_index]))
-        _fold_gate_at_index_in_moment(folded, moment_index, gate_index)
+        # Any moment with at least one gate is fair game
+        remaining_moment_indices = [i for i in remaining_gate_indices.keys() if remaining_gate_indices[i]]
```

Comment:
Rather than recalculating this every time by looping through all the keys, would it be more efficient to check in the last step if the last gate was removed and then to pop off the key?

Reply:
Yes! Great suggestion.